plugin/pulpcore/plugin/download/asyncio/factory.py

```diff
@@ -0,0 +1,116 @@
+import aiohttp
+import atexit
+from gettext import gettext as _
+import ssl
+from urllib.parse import urlparse
+
+from .http import HttpDownloader
+
+
+class DownloadFactory(object):
+    """
+    A factory for creating configured downloader objects that all share a single session.
+
+    Usage:
+        >>> import asyncio
+        >>> loop = asyncio._get_running_loop()
+        >>> the_factory = DownloadFactory(importer)
+        >>> downloader_a = the_factory.build(url_a)
+        >>> downloader_b = the_factory.build(url_b)
+        >>> downloads_list = [downloader_a, downloader_b]
+        >>> done, not_done = loop.run_until_complete(asyncio.wait(downloads_list))
+        >>> for task in done:
+        >>>     result = task.result()  # 'result' is a DownloadResult
+    """
+
+    def __init__(self, importer, download_class=HttpDownloader):
+        """
+
+        Args:
+            importer (pulpcore.plugin.models.Importer): The importer used to populate downloader
+                settings.
+            download_class (pulpcore.plugin.download.HttpDownloader): The asyncio-aware downloader
+                class this factory will use when instantiating downloaders. This option allows the
+                factory to work with custom downloaders. This is an optional parameter and defaults
+                to the HttpDownloader.
+        """
+        self._importer = importer
+        self._download_class = download_class
+        self._session = self._make_session_from_importer()
```

Comment:
So, every downloader created by this factory will share the same session?  So, the streamer will need to cache factories?

Reply:
The streamer will use `get_downloader()` right? If so, the importer instance will provide the session management. That way all the downloaders generated from the same importer instance will share their session. If this doesn't sound good, let me know.

Reply:
My concern is that the streamer will fetch the importer from the DB for each call.  So, if the importer stores the factory as an instance attribute, the factory will not be shared across streamer requests.  Unless I'm missing something.